src/inbox/mod_inbox_rdbms_async.erl

```diff
@@ -0,0 +1,276 @@
+
+-spec set_entry_properties(mongooseim:host_type(),
+                           mod_inbox:entry_key(),
+                           entry_properties()) ->
+    entry_properties() | {error, binary()}.
+set_entry_properties(HostType, Entry, Properties) ->
+    mod_inbox_rdbms:set_entry_properties(HostType, Entry, Properties).
+
+%% Internal functions
+
+flush_one(HostType, {set_inbox, Entry, Content, Count, MsgId, Timestamp}) ->
+    mod_inbox_rdbms:set_inbox(HostType, Entry, Content, Count, MsgId, Timestamp);
+flush_one(HostType, {set_inbox_incr_unread, Entry, Content, MsgId, Timestamp, Incrs}) ->
+    mod_inbox_rdbms:set_inbox_incr_unread(HostType, Entry, Content, MsgId, Timestamp, Incrs);
+flush_one(HostType, {remove_inbox_row, Entry}) ->
+    mod_inbox_rdbms:remove_inbox_row(HostType, Entry);
+flush_one(HostType, {reset_unread, Entry, MsgId}) ->
+    mod_inbox_rdbms:reset_unread(HostType, Entry, MsgId).
+
+-spec aggregate(NewTask :: async_task(), CurrentlyAccumulatedTask :: async_task()) -> FinalTask :: term().
+
+%%% First task being processed, just take that one
+aggregate(
+  Task,
+  undefined) ->
+    Task;
+
+%%% if new task is remove_inbox, ignore all previous requests and just remove
+aggregate(
+  {remove_inbox_row, Entry},
+  _) ->
+    {remove_inbox_row, Entry};
+
+%%% if the last task was remove_row, this task should now only be an insert
+aggregate(
+  {reset_unread, _, _},
+  {remove_inbox_row, _} = OldTask) ->
+    OldTask;
+aggregate(
+  {set_inbox, _, _, _, _, _} = NewTask,
+  {remove_inbox_row, _}) ->
+    NewTask;
+aggregate(
+  {set_inbox_incr_unread, _, _, _, _, _} = NewTask,
+  {remove_inbox_row, _}) ->
+    NewTask;
+
+%%% If the last task was a reset_unread,
+%   we prefer explicit resets
+%   then adhoc newer resets,
+%   then we accumulate inserts
+aggregate( %% an undefined means an explicit request to reset, it has priority
+  {reset_unread, _, undefined} = NewTask,
+  {reset_unread, _, _}) ->
+    NewTask;
+aggregate( %% an undefined means an explicit request to reset, it has priority
+  {reset_unread, _, _},
+  {reset_unread, _, undefined} = OldTask) ->
+    OldTask;
+aggregate( %% both are adhoc, we prefer the newer
+  {reset_unread, _, _} = NewTask,
+  {reset_unread, _, _}) ->
+    NewTask;
+aggregate(
+  {set_inbox, _, _, _, _, _} = NewTask,
+  {reset_unread, _, _}) ->
+    NewTask;
+aggregate( %% Here `Count` becomes an absolute value instead of an increment
+  {set_inbox_incr_unread, Entry, Content, MsgId, Timestamp, Incrs},
+  {reset_unread, _, _}) ->
+    {set_inbox, Entry, Content, Incrs, MsgId, Timestamp};
+
+%%% If the last task was a set_inbox
+aggregate( %% Reset refers to that same set_inbox
+  {reset_unread, _, MsgId},
+  {set_inbox, Entry, Content, _, MsgId, Timestamp}) ->
+    {set_inbox, Entry, Content, 0, MsgId, Timestamp};
+aggregate( %% Reset refers to some other set_inbox
+  {reset_unread, _, _},
+  {set_inbox, _, _, _, _, _} = OldTask) ->
+    OldTask;
```

Comment:
Why OldTask? As I understand it, someone wants to reset the inbox later, so why don't we set the count to zero?

Reply:
Ah, right, that seems like a lost counter. Should be set_inbox with a specific number, will fix.

Reply:
Ah, now I get it, this clause is correct. Look at the previous clause, it's matching the reset_unread with the set_inbox when they're having the same ID. If they're not having so, it means this is a reset_unread triggered by some different stanza, probably an older one, where the order was interleaved by two different clients of the same user, so we skip it.

What is indeed a bug is that this would eat explicit inbox resets, using the `entries` iqs. That one sets reset_unread with `undefined` as a message id, which simply resets the count to zero. I should match for that case.